klib/mbedtls.c

```diff
@@ -0,0 +1,301 @@
+        return MBEDTLS_ERR_SSL_WANT_READ;
+    len = MIN(buffer_length(conn->incoming), len);
+    kern_funcs.memcopy(buf, buffer_ref(conn->incoming, 0), len);
+    buffer_consume(conn->incoming, len);
+    return len;
+}
+
+static int tls_out_internal(tls_conn conn, buffer b)
+{
+    if (!b)
+        b = conn->outgoing;
+    int ret = mbedtls_ssl_write(&conn->ssl, buffer_ref(b, 0), buffer_length(b));
+    if (ret > 0) {
+        buffer_consume(b, ret);
+    } else if ((ret != MBEDTLS_ERR_SSL_WANT_READ) && (ret != MBEDTLS_ERR_SSL_WANT_WRITE)) {
+        tls_close(conn);
+        deallocate_buffer(b);
+        return -1;
+    }
+    if (buffer_length(b) == 0) {
+        deallocate_buffer(b);
+        if (b == conn->outgoing)
+            conn->outgoing = 0;
+    } else {
+        conn->outgoing = b;
+    }
+    return ret;
+}
+
+define_closure_function(1, 1, status, tls_out_handler,
+                        struct tls_conn *, conn,
+                        buffer, b)
+{
+    tls_conn conn = bound(conn);
+    if (b) {
+        tls_out_internal(conn, b);
+    } else {    /* application requested connection shutdown */
+        conn->app_in = 0; /* so that connection shutdown is not notified to application */
+        int ret = mbedtls_ssl_close_notify(&conn->ssl);
+        if ((ret == MBEDTLS_ERR_SSL_WANT_READ) || (ret == MBEDTLS_ERR_SSL_WANT_WRITE))
+            conn->state = tls_closing;
+        else
+            tls_close(conn);
+    }
+    return STATUS_OK;
+}
+
+define_closure_function(1, 1, status, tls_in_handler,
+                        struct tls_conn *, conn,
+                        buffer, b)
+{
+    tls_conn conn = bound(conn);
+    if (!b) {   /* underlying TCP connection closed */
+        conn->out = 0;
+        goto conn_close;
+    }
+    conn->incoming = b;
+    int ret;
+    switch (conn->state) {
+    case tls_handshake:
+        ret = mbedtls_ssl_handshake(&conn->ssl);
+        if (ret == 0) {
+            conn->state = tls_open;
+            conn->app_in = apply(conn->app_ch, init_closure(&conn->app_out, tls_out_handler, conn));
+            conn->app_ch = 0;   /* so that it is not invoked when the connection is closed */
+            if (!conn->app_in)  /* application-level error */
+                goto conn_close;
+        } else if ((ret != MBEDTLS_ERR_SSL_WANT_READ) && (ret != MBEDTLS_ERR_SSL_WANT_WRITE)) {
+            goto conn_close;
+        }
+        break;
+    case tls_open:
+        b = little_stack_buffer(512);
+        ret = mbedtls_ssl_read(&conn->ssl, buffer_ref(b, 0), b->length);
+        if (ret > 0) {
+            buffer_produce(b, ret);
+            apply(conn->app_in, b);
+        } else if ((ret != MBEDTLS_ERR_SSL_WANT_READ) && (ret != MBEDTLS_ERR_SSL_WANT_WRITE)) {
+            goto conn_close;
+        }
+        if (conn->outgoing && (tls_out_internal(conn, 0) < 0))
+            /* An error happened and the connection has been closed. */
+            return STATUS_OK;
```

Comment:
As things stand now, it's not possible to make an error status from a klib, because the timm* functions cannot be called: they are inline functions and contain a bunch of calls to kernel functions (such as intern, allocate_buffer, vbprintf, table_set), so if they are called from a klib they result in a linker error.

Reply:
This could be covered with a kernel-side helper to make the status ... even a result with just the integer return val would be better than no status.